File: client/sharedUI/src/commonMain/kotlin/com/example/client/RegisterView.kt
package com.example.client

import android.util.Log
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.client.register.RegisterClient
import com.example.client.register.RegisterRequestDto
import com.example.client.register.RegisterValidator
import dev.icerock.moko.resources.compose.stringResource
import kotlinx.coroutines.launch
import java.util.Locale
import kotlin.time.Duration.Companion.seconds

@Composable
@Preview
fun RegisterView() {
    var email by remember {mutableStateOf("")}
    var login by remember {mutableStateOf("")}
    var password by remember {mutableStateOf("")}
    var repeatPassword by remember {mutableStateOf("")}
    var errorKey by remember {mutableStateOf<String?>(null)}

    val code = remember {mutableStateOf("")}
    var codeSent:Boolean by remember {mutableStateOf(false)}
    var timerOn:Boolean by remember {mutableStateOf(false)}
    var resendCountdown by remember {mutableStateOf(60)}
    var confirmed:Boolean by remember {mutableStateOf(false)}

    val scope=rememberCoroutineScope()

    suspend fun validate():String?{
        val registerValidator=RegisterValidator()

        val loginValid=registerValidator.validateLogin(login)
        if(loginValid!=null) return loginValid

        val emailValid=registerValidator.validateEmail(email)
        if(emailValid!=null) return emailValid

        if(password!=repeatPassword) return "passwords_dont_match"

        val passwordValid=registerValidator.isPasswordValid(password)
        if(!passwordValid) return "invalid_password"

        return null
    }

    suspend fun register(){
        val language=Locale.getDefault().language.takeIf {it.isNotBlank()}?.uppercase()?:"EN"
        val registerRequestDto=RegisterRequestDto(login=login,displayName=login,email=email,password=password,language=language)
        val result=RegisterClient.register(registerRequestDto)
        if(result=="verification_code_sent"){
            errorKey=null
            codeSent=true
            timerOn=true
        }
        else if(result=="something_wrong"){
            errorKey="something_wrong"
        }
    }

    suspend fun confirmAccount(){
        val registerRequest=RegisterRequestDto(login=login,displayName=login,email=email,password=password)
        val confirmationResult=RegisterClient.registerConfirm(registerRequest,code.value)
        if(confirmationResult=="account_created"){
            confirmed=true
            timerOn=false
            codeSent=false
            errorKey=null
        }
        else{
            errorKey=confirmationResult
            code.value=""
        }
    }

    LaunchedEffect(timerOn){
        while(timerOn){
            if(resendCountdown>0){
                resendCountdown--
                kotlinx.coroutines.delay(1.seconds)
            }
            else{
                timerOn=false
            }
        }
    }

    MaterialTheme {
        Surface{
            Column(
                horizontalAlignment=Alignment.CenterHorizontally,
                modifier=Modifier.padding(top=25.dp).fillMaxWidth(),
                verticalArrangement=Arrangement.spacedBy(15.dp)
            ) {
                Text(text=stringResource(MR.strings.create_account),fontSize=28.sp,fontWeight=FontWeight.Bold)
                Text(text=stringResource(MR.strings.sign_up_to_continue),fontSize=20.sp)
                Spacer(modifier=Modifier.height(20.dp))
                TextField(
                    value=email,
                    onValueChange={email=it},
                    label={Text(stringResource(MR.strings.email_address))},
                    keyboardOptions=KeyboardOptions(
                        autoCorrectEnabled=false,
                        keyboardType=KeyboardType.Email,
                        imeAction=ImeAction.Next
                    )
                )
                TextField(
                    value=login,
                    onValueChange={login=it},
                    label={Text(stringResource(MR.strings.login))},
                    keyboardOptions=KeyboardOptions(
                        autoCorrectEnabled=false,
                        keyboardType=KeyboardType.Text,
                        imeAction=ImeAction.Next
                    )
                )
                TextField(
                    value=password,
                    onValueChange={password=it},
                    label={Text(stringResource(MR.strings.password))},
                    keyboardOptions=KeyboardOptions(
                        autoCorrectEnabled=false,
                        keyboardType=KeyboardType.Password,
                        imeAction=ImeAction.Next
                    ),
                    visualTransformation=PasswordVisualTransformation()
                )
                TextField(
                    value=repeatPassword,
                    onValueChange={repeatPassword=it},
                    label={Text(stringResource(MR.strings.re_enter_password))},
                    keyboardOptions=KeyboardOptions(
                        autoCorrectEnabled=false,
                        keyboardType=KeyboardType.Password,
                        imeAction=ImeAction.Done
                    ),
                    visualTransformation=PasswordVisualTransformation(),

                )
                errorKey?.let{key->
                    Text(text=stringResource(LocalText().getStringResource(key)),color=Color.Red)
                }
                Button(
                    onClick={
                        scope.launch{
                            val validationError:String?
                            try{
                                validationError=validate()
                            }
                            catch(e:Exception){
                                Log.e("validate register data",e.localizedMessage?:"")
                                return@launch
                            }
                            errorKey=validationError
                            if(validationError.isNullOrEmpty()){
                                try{
                                    register()
                                }
                                catch(e:Exception){
                                    Log.e("register",e.localizedMessage?:"")
                                    return@launch
                                }
                            }
                        }
                    }, enabled=!(email.isBlank()||login.isBlank()||password.isBlank()||repeatPassword.isBlank()
                                  ||confirmed)
                ){
                    Text(text=stringResource(MR.strings.sign_up))
                }

                if(codeSent){
                    Text(stringResource(MR.strings.verification_code_sent))
                    if(timerOn){
                        //todo decrease time
                        Text(text=stringResource(MR.strings.resend_code_in)+' '+resendCountdown)
                    }
                    Button(
                        onClick={
                            scope.launch{
                                val validationError:String?
                                try{
                                    validationError=validate()
                                }
                                catch(e:Exception){
                                    Log.e("validate register data",e.localizedMessage?:"")
                                    return@launch
                                }
                                errorKey=validationError
                                if(validationError.isNullOrEmpty()){
                                    try{
                                        register()
                                    }
                                    catch(e:Exception){
                                        Log.e("register",e.localizedMessage?:"")
                                        return@launch
                                    }
                                }
                            }
                        },
                        enabled=resendCountdown==0
                    ){
                        Text(stringResource(MR.strings.resend_code))
                    }

                    OtpInputField(
                        otp=code,
                        count=6,
                        otpBoxModifier=Modifier
                            .border(3.pxToDp(),Color.Black)
                            .background(Color.White),
                        otpTextType=KeyboardType.Number
                    )
                    Button(
                        onClick={
                            scope.launch{
                                try{
                                    confirmAccount()
                                }
                                catch(e:Exception){
                                    Log.e("register-confirm",e.localizedMessage?:"")
                                    return@launch
                                }
                            }
                        },
                        enabled=code.value.length==6
                    ){
                        Text(stringResource(MR.strings.confirm_account))
                    }
                }
                if(confirmed)
                    Text(stringResource(MR.strings.account_created))


                LabelledDivider(text=stringResource(MR.strings.or))
                Text(text=stringResource(MR.strings.special_sign_in_later))

            }
        }
    }
}

File: client/iosApp/iosApp/Account/Screens/RegisterView.swift
//
//  RegisterView.swift
//  iosApp
//
//  Created by Aleks Jankowiak on 16/09/2026.
//

import AuthenticationServices
import GoogleSignIn
import GoogleSignInSwift
import SharedLogic
import SwiftUI
import SwiftUIOTPEntry

struct RegisterView: View {
	let lang = Locale.current.language.languageCode ?? "EN"
	private enum Field: Int, CaseIterable {
		case email, login, password, confirmPassword
	}
	@State private var email: String = ""
	@State private var login: String = ""
	@State private var password: String = ""
	@State private var confirmPassword: String = ""
	@State private var code: String = ""

	@FocusState private var focusedField: Field?
	@State var isDismissKeyboard: Bool = false

	@State private var errorText: String? = ""
	@State private var codeSent: Bool = false
	@State private var timerOn: Bool = false
	@State private var resendCountdown: Int = 60
	@State private var confirmed: Bool = false

	let timer: Timer.TimerPublisher = Timer.publish(
		every: 1.0,
		on: .main,
		in: .common
	)

	let model: ModelUISwiftUIOTPEntry = .init(
		font: .systemFont(ofSize: 20),
		textAccessibilityForEmptyBox: "Empty box",
		textAccessibilityPosition: "Position",
		count: 6,
		spacing: 8,
		colorFocused: .red,
		colorEmpty: .gray,
		colorFill: .green,
		size: 50
	)

	var body: some View {
		NavigationStack {
			VStack {
				Text(localize(key: "create_account"))
					.font(.system(size: 28.0, weight: .bold))
				Text(localize(key: "sign_up_to_continue"))
					.font(.system(size: 20.0))
				Spacer()
				Form {
					TextField(
						localize(key: "email_address"),
						text: $email
					)
					.keyboardType(.emailAddress)
					.textContentType(.emailAddress)
					.font(.system(size: 24.0))
					.focused($focusedField, equals: .email)
					.autocorrectionDisabled()

					TextField(
						localize(key: "login"),
						text: $login
					)
					.textContentType(.username)
					.font(.system(size: 24.0))
					.focused($focusedField, equals: .login)
					.autocorrectionDisabled()

					SecureInputView(
						localize(key: "password"),
						text: $password
					)
					.textContentType(.password)
					.focused($focusedField, equals: .password)
					.font(.system(size: 24.0))

					SecureInputView(
						localize(key: "re_enter_password"),
						text: $confirmPassword
					)
					.textContentType(.password)
					.focused($focusedField, equals: .confirmPassword)
					.font(.system(size: 24.0))

				}.toolbar {
					ToolbarItem(placement: .keyboard) {
						Button(localize(key: "done")) {
							focusedField = nil
						}
					}
				}

				Text(errorText ?? "").foregroundStyle(.red)

				Button(action: validateFields) {
					Text(localize(key: "sign_up")).frame(maxWidth: .infinity)
				}
				.disabled(
					email.isEmpty || login.isEmpty || password.isEmpty
						|| confirmPassword.isEmpty
						|| (resendCountdown > 0 && codeSent) || confirmed
				)
				.padding()
				.background(Color.accentColor)
				.foregroundColor(.white)
				.glassEffect()
				.cornerRadius(25)
				.padding(.horizontal, 20)

				if codeSent {
					Text(localize(key: "verification_code_sent"))
					if timerOn {
						Text("\(localize(key: "resend_code_in")) \(resendCountdown)")
					}
					Button(action: validateFields) {
						Text(localize(key: "resend_code"))
					}
					.disabled(resendCountdown > 0)

					ViewSwiftUIOTPEntry(
						model: model,
						number: $code,
						isDismissKeyboard: $isDismissKeyboard
					)
					Button(action: confirmAccount) {
						Text(localize(key: "confirm_account"))
					}
					.buttonStyle(.glass)
					.disabled(code.count < 6)
				}
				if confirmed {
					Text(localize(key: "account_created"))
				}

				LabelledDivider(label: localize(key: "or"))
				//			GoogleSignInButton(
				//				scheme: .light,
				//				state: .normal,
				//				action: {handleGoogleSignInButton()})
				//			.frame(width: 280, height: 45)
				//
				//			SignInWithAppleButton(.continue){
				//				request in request.requestedScopes=[.email]
				//			} onCompletion: { result in
				//				switch result {
				//					case .success(let authorization):
				//						print("\(localize(key: "authorization_successful")): \(authorization)")
				//					case .failure(let error):
				//						print("\(localize(key: "authorization_failed"))): \(error.localizedDescription)")
				//				}
				//			}.frame(width: 280, height: 45)

				//Text(localize(key: "special_sign_in_later"))
				//Spacer()
				HStack {
					Text(localize(key: "already_have_account"))
					NavigationLink(destination: SignInView()) {
						Text(localize(key: "sign_in"))
					}
				}

			}
			.padding(.top, 15)
			.onReceive(
				timer.autoconnect(),
				perform: { _ in
					if timerOn {
						if resendCountdown > 0 {
							resendCountdown = resendCountdown - 1
						} else {
							timerOn = false
						}
					}
				}
			)
		}
	}

	func validateFields() {
		Task {
			let registerValidator = RegisterValidator()
			do {
				let loginValid = try await registerValidator.validateLogin(
					login: login
				)
				if loginValid != nil {
					errorText = localize(key: loginValid!)
					return
				}

				let emailValid = try await registerValidator.validateEmail(
					email: email
				)
				if emailValid != nil {
					errorText = localize(key: emailValid!)
					return
				}

				if password != confirmPassword {
					errorText = localize(key: "passwords_dont_match")
					return
				}

				let passwordValid = registerValidator.isPasswordValid(
					password: password
				)
				if passwordValid != true {
					errorText = localize(key: "invalid_password")
				}
				errorText = ""
			} catch let error {
				print(error)
				return
			}
			createAccount()
		}
	}

	func createAccount() {
		Task {
			do {
				let langCode: String = lang.identifier.uppercased()
				let registerRequest = RegisterRequestDto(
					login: login,
					displayName: login,
					email: email,
					password: password,
					language: langCode
				)
				let result = try await RegisterClient().register(
					request: registerRequest
				)
				if result == "verification_code_sent" {
					errorText = ""
					codeSent = true
					timerOn = true
				} else if result == "something_wrong" {
					errorText = localize(key: "something_wrong")
				}

			} catch let error {
				print(error)
				return
			}
		}
	}

	func confirmAccount() {
		Task {
			do {
				let langCode: String = lang.identifier.uppercased()
				let registerRequest = RegisterRequestDto(
					login: login,
					displayName: login,
					email: email,
					password: password,
					language: langCode
				)
				let confirmationResult = try await RegisterClient().registerConfirm(
					request: registerRequest,
					code: code
				)
				print(confirmationResult)
				if confirmationResult == "account_created" {
					confirmed = true
					timerOn = false
					codeSent = false
				} else {
					errorText = localize(key: confirmationResult)
				}

			} catch let error {
				print(error)
				return
			}
		}
	}

	//	func handleGoogleSignInButton() {
	//		guard let rootViewController = UIApplication.shared.rootViewController else {
	//			// Handle error
	//			return
	//		}
	//
	//		GIDSignIn.sharedInstance.signIn(withPresenting: rootViewController) { signInResult, error in
	//			guard error == nil else { print(error ?? "none");return}
	//			guard let signInResult = signInResult else { return }
	//
	//			let user = signInResult.user
	//			//let emailAddress = user.profile?.email
	//			signInResult.user.refreshTokensIfNeeded { user, error in
	//				guard error == nil else { return }
	//				guard let user = user else { return }
	//
	//				let idToken = user.idToken
	//				print(idToken)
	//				// Send ID token to backend (example below).
	//			}
	//		}
	//	}

}

#Preview {
	RegisterView().preferredColorScheme(.dark)
}
